File: tests/run_conformance_cases.mjs
// Run the shared adversarial corpus against a JavaScript verifier.
//
//     node tests/run_conformance_cases.mjs [--json] [--browser]
//
// Default target is the Node reference verifier (er1_verify.mjs); --browser runs the SAME corpus
// against the browser build (verify/er1_verify.browser.mjs) under Node's own WebCrypto. The
// corpus also runs against Python (tests/run_conformance_cases.py). Three implementations, one
// set of expectations — a defect fixed in one language cannot silently survive in another.
// tests/test_conformance_corpus.py is what makes all three gate CI.
import { readFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

const here = dirname(fileURLToPath(import.meta.url));
const useBrowser = process.argv.includes("--browser");
const modulePath = useBrowser
  ? join(here, "..", "verify", "er1_verify.browser.mjs")
  : join(here, "..", "er1_verify.mjs");
const implementation = useBrowser ? "er1_verify.browser.mjs" : "er1_verify.mjs";
// The browser build's verify() is async (WebCrypto); awaiting a sync return is harmless.
const { verify } = await import(modulePath);
const corpus = JSON.parse(readFileSync(join(here, "conformance_cases.json"), "utf8"));

const asJson = process.argv.includes("--json");
const results = [];
let failed = 0;

for (const c of corpus.cases) {
  let res;
  try {
    res = await verify(c.doc);
  } catch (exc) {
    res = { ok: null, errors: [`THREW: ${exc.message}`] };   // a crash is always a failure
  }
  const errs = (res.errors ?? []).join(" | ");
  const okMatch = res.ok === c.expect_ok;
  const errMatch = !c.error_contains || errs.includes(c.error_contains);
  const pass = okMatch && errMatch;
  if (!pass) failed++;
  results.push({ name: c.name, pass, ok: res.ok, errors: errs });
  if (!asJson) {
    console.log(`${pass ? "ok  " : "FAIL"} ${c.name}`);
    if (!pass) {
      console.log(`      expected ok=${c.expect_ok} error~${JSON.stringify(c.error_contains)}`);
      console.log(`      got      ok=${res.ok} errors=${JSON.stringify(errs)}`);
    }
  }
}

if (asJson) {
  console.log(JSON.stringify({ implementation, failed, results }, null, 1));
} else {
  console.log(`\n${corpus.cases.length - failed}/${corpus.cases.length} conformance cases pass`);
}
process.exit(failed ? 1 : 0);

File: tests/run_conformance_cases.py
#!/usr/bin/env python3
"""Run the shared adversarial corpus against the Python reference verifier.

    python3 tests/run_conformance_cases.py

Also exposed as pytest cases so `pytest` covers it. The same corpus runs against Node
(tests/run_conformance_cases.mjs) and the browser build — three implementations, one set of
expectations.
"""
import json
import pathlib
import sys

HERE = pathlib.Path(__file__).parent
sys.path.insert(0, str(HERE.parent))

import er1_verify as ER1  # noqa: E402

CORPUS = json.loads((HERE / "conformance_cases.json").read_text())


def run_case(c: dict) -> tuple[bool, str]:
    try:
        res = ER1.verify(c["doc"])
    except Exception as exc:                      # a crash is always a failure
        return False, f"THREW: {type(exc).__name__}: {exc}"
    errs = " | ".join(res["errors"])
    if res["ok"] != c["expect_ok"]:
        return False, f"ok={res['ok']} (expected {c['expect_ok']}) errors={errs}"
    if c["error_contains"] and c["error_contains"] not in errs:
        return False, f"missing {c['error_contains']!r} in errors={errs}"
    return True, errs


def test_every_conformance_case():
    failures = []
    for c in CORPUS["cases"]:
        ok, detail = run_case(c)
        if not ok:
            failures.append(f"{c['name']}: {detail}")
    assert not failures, "\n".join(failures)


if __name__ == "__main__":
    bad = 0
    for case in CORPUS["cases"]:
        ok, detail = run_case(case)
        bad += not ok
        print(f"{'ok  ' if ok else 'FAIL'} {case['name']}")
        if not ok:
            print(f"      {detail}")
    total = len(CORPUS["cases"])
    print(f"\n{total - bad}/{total} conformance cases pass")
    raise SystemExit(1 if bad else 0)
